### PR TITLE
REFACTOR: Some component and move constants to constants directory

### PR DESCRIPTION
**Pull request by sahsisunny**
Date: 8 November 2023
Developer Name: @sahsisunny 

----
## Issue Ticket Number:- 
-  #19 

## Description: 
This pull request introduces significant enhancements and optimizations to the codebase, focusing on modularity, styling, and code organization. Here's a detailed breakdown of the changes:

Is Under Feature Flag 
- [ ] Yes
- [x] No

Database changes
- [ ] Yes
- [x] No

Breaking changes (If your feature is breaking/missing something please mention pending tickets)
- [ ] Yes
- [x] No

### Is Development Tested?

- [x] Yes
- [ ] No

#### Added Components
- **App Page Component**
  - `src/components/App/InputSection.tsx` : New component for input section
  - `src/components/App/OutputSection.tsx` : New component for output section

- **Footer Component**
  - `src/components/Footer/index.tsx` : Background color change

- **Layout Component**
  - `src/components/Layout/index.tsx` : Removed unnecessary CSS

- **ProfileIcon Component**
  - `src/components/ProfileIcon/ProfileIcon.tsx` : CSS changes to use Tailwind CSS

- **Redirect Page Component**
  - `src/components/Redirect/LoaderTimer.tsx` : New component for loader timer
  - `src/components/Redirect/NotFound.tsx` : New component for not found page
  - `src/components/Redirect/RedirectFooter.tsx` : New component for redirect page

#### Constants
- `src/constants/constants.ts` : Moved constants from utils to constants directory, added comments for regex

#### Page Components
- `src/pages/[redirect]/index.tsx` : Separated code for redirect page into different components
- `src/pages/app/index.tsx` : Separated code for app page into different components
- `src/pages/login/index.tsx` : Changed import paths for constants moved to constants directory

#### Global Styles
- `src/styles/global.css` : Modified h-container CSS

#### Login Modal Component
- `src/components/LoginModal/index.tsx` : Use dialog element instead of div

#### Demo
https://github.com/Real-Dev-Squad/tiny-site-frontend/assets/70854507/092fff7c-45ad-450e-990b-a25870cf5b73

#### Tests Added
File                        | % Stmts | % Branch | % Funcs | % Lines | Uncovered Line #s       
----------------------------|---------|----------|---------|---------|-------------------------
All files                   |   82.45 |    63.95 |   71.15 |   82.11 |                         
 src/components/App|                        
  InputSection.tsx          |     100 |      100 |     100 | 100 |                                                                            
  OutputSection.tsx         |     100 |      100 |     100 |     100 |                                                                            
 src/components/Footer/index.tsx                 |     100 |      100 |     100 |     100 | 
 src/components/Layout/index.tsx                 |     100 |      100 |     100 |     100 | 
 src/components/LoginModal/index.tsx                 |     100 |      100 |     100 |     100 | 
 src/components/Navbar/index.tsx                 |   94.44 |    66.66 |      75 |   94.11 | 33
 src/components/ProfileIcon/ProfileIcon.tsx           |     100 |       50 |     100 |     100 | 4
 src/components/Redirect|
  LoaderTimer.tsx           |     100 |      100 |     100 |     100 | 
  NotFound.tsx              |     100 |      100 |     100 |     100 | 
  RedirectFooter.tsx        |     100 |      100 |     100 |     100 | 
 src/pages/[redirect]/index.tsx               | 65.85 |    27.27 |   42.85 |   68.42 | 29-30,35-36,41-45,50-53
 src/pages/app/index.tsx                 |   57.14 |    33.33 |   42.85 |   56.25 | 23-28,38-43,48-51,57-85

# Note for Reviewers

This pull request focuses on the refactoring of existing components and does not currently include tests for API calls. To enhance the test coverage and ensure comprehensive testing of API interactions, a subsequent pull request will be submitted after the implementation of MSW (Mock Service Worker).

The introduction of MSW will allow us to mock API calls during testing, providing a controlled environment for testing various scenarios and ensuring the stability and reliability of our components.

The upcoming PR will specifically address the testing of API interactions, and we aim to maintain a high standard of test coverage for both positive and edge cases.

Thank you for your understanding, and any feedback on the current refactoring work is highly appreciated.

### To-Do in Upcoming PR:
- Implement MSW for API mocking.
- Add tests for API interactions.
- Ensure comprehensive coverage for various scenarios.
- Maintain high-quality test standards.

### CODE REVIEW

File: src/pages/[redirect]/index.tsx [hunk 1]
```diff
@@ -1,85 +1,50 @@
...
-);
-
 const Redirect = () => {
     const router = useRouter();
     const { redirect: shortUrlCode } = router.query as { redirect: string };
     const [originalUrl, setOriginalUrl] = useState('');
     const [timer, setTimer] = useState(5);
     const [isPremiumUser] = useState(false);
     const [showTooltip, setShowTooltip] = useState(false);
+    const [notFound, setNotFound] = useState(false);
```
Comment (vinit717) [inline src/pages/[redirect]/index.tsx:17]: Can you decrease the number of usestate

File: src/pages/app/index.tsx [hunk 2]
```diff
@@ -1,81 +1,21 @@
...
-);
-
 const App = () => {
     const [url, setUrl] = useState<string>('');
     const [shortUrl, setShortUrl] = useState<string>('');
     const [toastMessage, setToastMessage] = useState<string>('');
     const [showToast, setShowToast] = useState<boolean>(false);
-    const [showInputBox, setShowInputBox] = useState<boolean>(false);
+    const [showInputBox, setShowInputBox] = useState<boolean>(true);
+    const [showOutputBox, setShowOutputBox] = useState<boolean>(false);
```
Comment (vinit717) [inline src/pages/app/index.tsx:18]: here also

File: src/components/Redirect/RedirectFooter.tsx [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import Link from 'next/link';
+
```
Comment (vinit717) [inline src/components/Redirect/RedirectFooter.tsx:2]: why a blank line?

File: src/pages/[redirect]/index.tsx (hunk 1 above)
Comment (sahsisunny) [reply src/pages/[redirect]/index.tsx:17]: Can you suggest to me how to decrease it?

File: src/pages/app/index.tsx (hunk 2 above)
Comment (sahsisunny) [reply src/pages/app/index.tsx:18]: Using the "Add a Suggestion" GitHub feature to suggest to me how to reduce use useState please

File: src/pages/[redirect]/index.tsx (hunk 1 above)
Comment (sahsisunny) [reply src/pages/[redirect]/index.tsx]: removed isPremiumUser useState

Comment (vinit717) [reply src/pages/[redirect]/index.tsx]: either consolidates the component's state into a single state object, or currently I am looking into state management library which one to choose I have Zotai, Zustand, SWR, Redux, Recoil in mind

File: src/pages/app/index.tsx (hunk 2 above)
Comment (vinit717) [reply src/pages/app/index.tsx:18]: const [state, setState] = useState({
        url: '',
        shortUrl: '',
        toastMessage: '',
        showToast: false,
        showInputBox: true,
        showOutputBox: false,
    });
    
    something like this

File: src/components/Redirect/RedirectFooter.tsx (hunk 3 above)
Comment (vinit717) [reply src/components/Redirect/RedirectFooter.tsx:2]: but there should be no blank line in middle of imports statement

File: src/pages/app/index.tsx (hunk 2 above)
Comment (sahsisunny) [reply src/pages/app/index.tsx:18]: I believe using multiple useState hooks is more readable than what you suggested.

File: src/components/Redirect/RedirectFooter.tsx (hunk 3 above)
Comment (sahsisunny) [reply src/components/Redirect/RedirectFooter.tsx:2]: Yes, it separates the import based on the imported categories like external library, components, styles, etc.

File: src/pages/app/index.tsx (hunk 2 above)
Comment (sahsisunny) [reply src/pages/app/index.tsx:18]: It is perfectly fine to use multiple useState hooks as needed.

File: src/pages/[redirect]/index.tsx (hunk 1 above)
Comment (sahsisunny) [reply src/pages/[redirect]/index.tsx]: This conversion should be resolved, and we can discuss which library to use on a call.

File: src/components/App/OutputSection.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            {originalUrl}
+        </span>
+        <AiOutlineArrowDown style={{ fontSize: '3rem' }} />
+
+        <div
+            className="text-white flex flex-col md:flex-row justify-center items-center space-y-2 sm:space-y-0 space-x-0 sm:space-x-2 rounded-2xl w-full  lg:w-[80%] xl:w-[80%] cursor-pointer bg-black p-1"
+            onClick={handleCopyUrl}
+        >
+            <span className="ml-2 p-4 text-center w-full sm:w-[80%] sm:text-2xl md:text-3xl xl:text-3xl">
+                {shortUrl.replace(/(^\w+:|^)\/\//, '')}
```
Comment (vinit717) [inline src/components/App/OutputSection.tsx:32]: can you move this to constant

Comment (sahsisunny) [reply src/components/App/OutputSection.tsx:32]: No, we cannot move this to constants. Since 'replace' is the function, creating a separate utils function is not a good idea. Therefore, we can keep this as is.

File: __tests__/pages/[redirect].test.tsx [hunk 5]
```diff
@@ -20,30 +20,31 @@ describe('Redirect Component', () => {
...
-        await act(async () => {
-            fireEvent.click(goButton);
-        });
-        expect(mockRouterPush).toHaveBeenCalled();
+        expect(timer).toBeInTheDocument();
+        expect(goButton).toBeInTheDocument();
     });
 
-    test('show tooltip on Go button click', async () => {
+    test.skip('redirects to original URL on Go button click', async () => {
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply __tests__/pages/[redirect].test.tsx:31]: As mentioned in the description, we cannot test any API calls, so I am setting up MSW to perform the testing.

File: src/components/Redirect/LoaderTimer.tsx [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+const LoaderTimer = ({ timer, goButtonClickHandler }: LoaderTimerProps) => {
+    if (timer < 1) {
+        return (
+            <div className="mt-4 flex flex-col items-center space-y-2">
+                <RedirectIcon />
+                <p className="text-1xl">Redirecting...</p>
+            </div>
+        );
+    } else {
```
Comment (Ajeyakrishna-k) [inline src/components/Redirect/LoaderTimer.tsx:18]: Do we need this else block?

**Review by Ajeyakrishna-k — COMMENTED** [review]
Please also remove unnecessary spaces between import statements and classnames. There are a lot of them.

File: src/components/App/OutputSection.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+            {originalUrl}
+        </span>
+        <AiOutlineArrowDown style={{ fontSize: '5rem', padding: '15px' }} />
+
+        <div
+            className="text-white flex flex-col md:flex-row justify-center items-center  rounded-2xl w-full  lg:w-[80%] xl:w-[80%] cursor-pointer bg-black p-2"
+            onClick={handleCopyUrl}
+        >
+            <span className="ml-2 p-4 text-center w-full sm:w-[80%] ellipsis overflow-hidden overflow-ellipsis whitespace-nowrap sm:text-2xl md:text-3xl xl:text-3xl">
+                {shortUrl.replace(/(^\w+:|^)\/\//, '')}
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply src/components/App/OutputSection.tsx:32]: https://github.com/Real-Dev-Squad/tiny-site-frontend/pull/53#discussion_r1393465321

**Comment by sahsisunny** [comment]
> Please also remove unnecessary spaces between import statements and classnames. There are a lot of them. 

https://github.com/Real-Dev-Squad/tiny-site-frontend/pull/53#discussion_r1392274577

File: src/components/App/InputSection.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import { ChangeEvent, FormEvent } from 'react';
+
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply src/components/App/InputSection.tsx:2]: https://github.com/Real-Dev-Squad/tiny-site-frontend/pull/53#discussion_r1392274577

File: src/components/Redirect/LoaderTimer.tsx (hunk 6 above)
Comment (sahsisunny) [reply src/components/Redirect/LoaderTimer.tsx:18]: Both returns are in the same block so we need this else block

File: src/components/App/OutputSection.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                </Link>
+                <Button
+                    type="button"
+                    className="bg-gray-900 md:rounded-r-2xl md:py-5 md:px-4 hover:bg-gray-800  w-full sm:w-[50%] rounded-r-2xl md:rounded-none flex justify-center items-center after:content-['Copy'] md:after:content-[''] sm:px-2 sm:py-1 py-1"
+                    testId="copy-button"
+                    onClick={handleCopyUrl}
+                >
+                    <IoIosCopy style={{ fontSize: '1.5rem' }} />
+                    &nbsp;
+                </Button>
```
[the remark this answers is not in the document]
Comment (sahsisunny) [reply src/components/App/OutputSection.tsx:54]: Why Do we need to create separate components for the copy button?

File: src/components/Redirect/LoaderTimer.tsx (hunk 6 above)
Comment (Ajeyakrishna-k) [reply src/components/Redirect/LoaderTimer.tsx:18]: You are **returning** when timer is less than 1. So if you remove the else block you will only remove a level of nesting. Functionality will remain same.

File: src/components/App/OutputSection.tsx (hunk 9 above)
[the remark this answers is not in the document]
Comment (Ajeyakrishna-k) [reply src/components/App/OutputSection.tsx:54]: Link and button with its functionality.

File: src/components/App/OutputSection.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (Ajeyakrishna-k) [reply src/components/App/OutputSection.tsx:32]:
```suggestion
                const pattern = /(^\w+:|^)\/\//;
                {shortUrl.replace( pattern,'')}
```

File: src/components/Redirect/LoaderTimer.tsx (hunk 6 above)
Comment (sahsisunny) [reply src/components/Redirect/LoaderTimer.tsx:18]: Done

File: src/components/App/OutputSection.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (sahsisunny) [reply src/components/App/OutputSection.tsx]: Done

**Review by Ajeyakrishna-k — APPROVED** [review]

**Review by vinit717 — APPROVED** [review]
